### DISCUSSION TITLE
Python package for ETAPI

### DISCUSSION THREAD

**Discussion by Nriver** [TriliumNext/Trilium#2669] [Show and tell]
I made a python package called `trilium-py` for Trilium Note's ETAPI.

I have uploaded it to [PYPI](https://pypi.org/project/trilium-py/), so anyone who is interested can install it with pip.

Source code is [here](https://github.com/Nriver/trilium-py). There are some simple demo code in the document. You are welcomed to try it.

With the power of python, there are a lot more possibilities for Trilium.

**Comment by Nriver** [comment]
I added functions `get_todo`, `todo_check`, `todo_uncheck` to manipulate today's todo list. You can get todo list from trilium and check or uncheck them with program now!

**Comment by zadam** [comment]
Nice! You can put the link to ETAPI wiki if you want.

**Reply by Nriver** [reply]
> Nice! You can put the link to ETAPI wiki if you want.

Thanks! I just added it to wiki.

**Comment by sigaloid** [comment]
Awesome! I'm in the early stages of making a Rust library so this is helpful :smile:

**Comment by Nriver** [comment]
New features:
`add_todo` to add todo item
`update_todo` to update a existing todo's description text
`delete_todo` to delete a todo item from the list

**Comment by Nriver** [comment]
I just made a new feature: move yesterday's unfinished/unchecked todo item to today's note, you can use it by calling `move_yesterday_unfinished_todo_to_today`.

It can be used in combine with scheduled tasks. e.g. you can make it run everyday at 00:01 to move what you haven't done yesterday to today. I find it works well with [schedule](https://github.com/dbader/schedule).

**Comment by Nriver** [comment]
Recent updates:
1. add import markdown notes, the program should handle the image import for you. Tested with [VNote](https://github.com/vnotex/vnote) and Logseq.
2. moving TODO items will keep the internal and external links, text styles now.

I've created a repo https://github.com/Nriver/awesome-trilium to keep a record of awesome trilium notes related creations. If you create anything cool with trilium-py, do not hesitate to make PR and share with us.

**Comment by Nriver** [comment]
For https://github.com/zadam/trilium/issues/2259#issuecomment-1454987523, I've updated the steps to import note from Obsidian.

https://github.com/Nriver/trilium-py#import-from-obsidian

**Comment by Nriver** [comment]
Add support for creating backups via ETAPI https://github.com/Nriver/trilium-py#-create-data-backup This should be supported in the latest Trilium 0.60.3

Also add an example for import/migrate note from Joplin to Trilium. https://github.com/Nriver/trilium-py#import-from-joplin

**Comment by Nriver** [comment]
For https://github.com/zadam/trilium/discussions/4056, I've made a script to convert Turtl's json data into ordinary markdown file [here](https://github.com/Nriver/trilium-py/tree/main/examples/turtl-to-markdown).

Also implemented full featured file-upload in markdown import. Link to local files should be uploaded to Trilium without problem. [Import from Turtl](https://github.com/Nriver/trilium-py/tree/main#import-from-turtl)
